security/pkg/pki/ca/ca.go

```diff
@@ -157,51 +137,62 @@ func NewSelfSignedIstioCAOptions(ctx context.Context,
-			Org:          org,
-			IsCA:         true,
-			IsSelfSigned: true,
-			RSAKeySize:   caRSAKeySize,
-			IsDualUse:    dualUse,
-		}
-		pemCert, pemKey, ckErr := util.GenCertKeyFromOptions(options)
-		if ckErr != nil {
-			return nil, fmt.Errorf("unable to generate CA cert and key for self-signed CA (%v)", ckErr)
+	b := backoff.NewExponentialBackOff(backoff.DefaultOption())
+	err = b.RetryWithContext(ctx, func() error {
+		// For the first time the CA is up, if readSigningCertOnly is unset,
+		// it generates a self-signed key/cert pair and write it to CASecret.
+		// For subsequent restart, CA will reads key/cert from CASecret.
+		caSecret, err := client.Secrets(namespace).Get(context.TODO(), CASecret, metav1.GetOptions{})
+		if err == nil {
+			pkiCaLog.Infof("Load signing key and cert from existing secret %s:%s", caSecret.Namespace, caSecret.Name)
+			rootCerts, err := util.AppendRootCerts(caSecret.Data[CACertFile], rootCertFile)
+			if err != nil {
+				return fmt.Errorf("failed to append root certificates (%v)", err)
+			}
+			if caOpts.KeyCertBundle, err = util.NewVerifiedKeyCertBundleFromPem(caSecret.Data[CACertFile],
+				caSecret.Data[CAPrivateKeyFile], nil, rootCerts); err != nil {
+				return fmt.Errorf("failed to create CA KeyCertBundle (%v)", err)
+			}
+			pkiCaLog.Infof("Using existing public key: %v", string(rootCerts))
+			return nil
 		}
+		if apierror.IsNotFound(err) {
+			pkiCaLog.Infof("CASecret %s not found, will create one", CASecret)
+			options := util.CertOptions{
+				TTL:          caCertTTL,
+				Org:          org,
+				IsCA:         true,
+				IsSelfSigned: true,
+				RSAKeySize:   caRSAKeySize,
+				IsDualUse:    dualUse,
+			}
+			pemCert, pemKey, ckErr := util.GenCertKeyFromOptions(options)
+			if ckErr != nil {
+				return fmt.Errorf("unable to generate CA cert and key for self-signed CA (%v)", ckErr)
+			}
 
-		rootCerts, err := util.AppendRootCerts(pemCert, rootCertFile)
-		if err != nil {
-			return nil, fmt.Errorf("failed to append root certificates (%v)", err)
-		}
+			rootCerts, err := util.AppendRootCerts(pemCert, rootCertFile)
+			if err != nil {
+				return fmt.Errorf("failed to append root certificates (%v)", err)
+			}
+			if caOpts.KeyCertBundle, err = util.NewVerifiedKeyCertBundleFromPem(pemCert, pemKey, nil, rootCerts); err != nil {
+				return fmt.Errorf("failed to create CA KeyCertBundle (%v)", err)
+			}
 
-		if caOpts.KeyCertBundle, err = util.NewVerifiedKeyCertBundleFromPem(pemCert, pemKey, nil, rootCerts); err != nil {
-			return nil, fmt.Errorf("failed to create CA KeyCertBundle (%v)", err)
+			// Write the key/cert back to secret so they will be persistent when CA restarts.
+			secret := k8ssecret.BuildSecret(CASecret, namespace, nil, nil, nil, pemCert, pemKey, istioCASecretType)
+			if _, err = client.Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{}); err != nil {
+				pkiCaLog.Errorf("Failed to write secret to CA (error: %s). Abort.", err)
```

Comment:
`Abort` is not accurate any more, right? Now we retry?

Reply:
very nit: may be nice to specifically handle the IsAlreadyExist error with a non `Errorf` level log? Since it is expected case

Reply:
good catch